### ISSUE TITLE
Ability to Merge LORA files into main model files

### ISSUE THREAD

**Issue by ollobrains** [easydiffusion/easydiffusion#1104]
none of the other programs i have come across for stable diffusion have the ability to merge LORA files directly into a model, expanding its abilities with very specific additions. Both are safe tensors, it would be a great feature if it was easy to do and use. Currently we get a tensor a doesnt match tensor b size error but a script i did find which is difficult to use.

If something similar that worked could be put into model merging / beta LORA builds it might attract a lot more users, automatic1111 and others do not have it yet.

https://github.com/kohya-ss/sd-scripts this project does it but i cant get my head around it, a much simplier way must exist.

Thoughts , im sure many more users would get involved, its a much asked for feature on reddit etc, but no one has done it yet

**Comment by ollobrains** [comment]
another dev has come up with this as well that might assist with implementation https://github.com/DevArqSangoi/das-dataAugmentation

**Comment by DevArqSangoi** [comment]
I don't think LoRAs are supposed to do this kind of thing. They are just a fine-tune that you can add to any model with a trigger. It's even better than having it merged into one specific model because you can use it with any of them.

**Comment by ollobrains** [comment]
> I don't think LoRAs are supposed to do this kind of thing. They are just a fine-tune that you can add to any model with a trigger. It's even better than having it merged into one specific model because you can use it with any of them.

actually LORAs are supposed to do this, if u look at civitai.com u will see hundreds of fine tuned models, usually 50-100 LORAs merged into a main model usually around enhancing certain themes like fantasy, RPG that sort of thing, and they are quite well recieved.  Its an easy way to add new triggers where stable diffusion lacks them, its even better than using stand alone merges for quite a few people. 

So there are lots of people who disagree 100% with the viewpoint presented

**Comment by ollobrains** [comment]
> I don't think LoRAs are supposed to do this kind of thing. They are just a fine-tune that you can add to any model with a trigger. It's even better than having it merged into one specific model because you can use it with any of them.

Ok let me rephrase from a coding perspective if u wanted to add it, how difficult would it be to make it happen .  taking out any pre concieved personal biases as to what LORAs are and arent meant to do (given they are safetensors to) the technical implementation and coding time is it easy or hard to do ?

**Comment by uwidev** [comment]
> I don't think LoRAs are supposed to do this kind of thing. They are just a fine-tune that you can add to any model with a trigger. It's even better than having it merged into one specific model because you can use it with any of them.

LoRA's inserts itself into the model and nudges generation towards specific features. But it can only nudge so much, where when given a vastly different model, it can't nudge enough to get desired results. That said, the LoRA can still work with similar descendants, or perhaps even the parent model, but it will still have diminishing results.

Having a built-in way to merge the LoRA into the model would be interesting. I've haven't yet directly worked on models, let alone merging, but at a glance, it seems like a way to extend a model as a complete package. 

I think it's a good idea, but I think there might be some problems about compatibility and usability. Say, someone wants a certain LoRA, but it's embedded in a model. A user would then need to download the entire model rather than the much smaller LoRA. While this is trivially solved by having the user upload both, I think it's something to consider.

**Comment by heleusbrands** [comment]
To answer ollobrains question - from a programming perspective it's not hard at all if you're familiar with the stable diffusion side of ML modeling and training. Although, it might be more of a question of 'how' do you want to go about it,  since usually the best way to achieve this is through fine-tuning an existing model (possibly with a higher learning rate, depending on the data). 

And in actuality, you're both sorta right. Loras weren't _designed_ with that purpose in mind, since programmers can just fine tune the model and achieve better results. As I understand it, they were designed to be more of a convenient stylistic addon to extend a models existing functionality temporarily for a specific set of tasks. There are many times that you might not want to alter the base model through further training, but the people using your model very well might. That's kinda where Loras come in.  It prevents the need to finetune and create an entire new model just for a few stylistic additions. 

However, just because they weren't designed for that purpose, doesn't mean they can't be utilized that way, especially since fine-tuning a sd model generally requires that the diffusion artist also be a programmer, which is less and less the case.

Which brings me to the main question: "Can they be easily merged?". 

As a programmer, absolutely! 

As an sd prompt artist - I think so. I seem to remember testing out a merging feature in either Automatic1111 or in EasyDiffusion (I can't remember which atm), and wanted to see what would happen if I merged a Lora safetensors with a an sd 1.5 based safetensors model. And believe it or not, it actually worked without a hitch! Shouldn't have surprised me, considering they're both very similar in design, but I definitely was half expecting it to throw and error there. Lol

So all that to say, give it a try in one, and if that doesn't work - try it in the other! :) @ollobrains

**Comment by JeLuF** [comment]
EasyDiffusion 3.0, released today, comes with LORA support.
